ingest/ledgerbackend/buffered_storage_backend.go

```
@@ -0,0 +1,304 @@
+		if sequence < bsb.ledgerMetaArchive.GetStartLedgerSequence() {
+			return errors.New("requested sequence preceeds current LedgerCloseMetaBatch")
+		}
+
+		// Sequence is beyond the current LedgerCloseMetaBatch
+		lcmBatchBinary, err := bsb.ledgerBuffer.getFromLedgerQueue()
+		if err != nil {
+			return errors.Wrap(err, "failed getting next ledger batch from queue")
+		}
+
+		// Turn binary into xdr
+		err = bsb.ledgerMetaArchive.Data.UnmarshalBinary(lcmBatchBinary)
+		if err != nil {
+			return errors.Wrap(err, "failed unmarshalling lcmBatchBinary")
+		}
+	}
+}
+
+// nextExpectedSequence returns nextLedger (if currently set) or start of
+// prepared range. Otherwise it returns 0.
+// This is done because `nextLedger` is 0 between the moment Stellar-Core is
+// started and streaming the first ledger (in such case we return first ledger
+// in requested range).
+func (bsb *BufferedStorageBackend) nextExpectedSequence() uint32 {
+	if bsb.nextLedger == 0 && bsb.prepared != nil {
+		return bsb.prepared.from
+	}
+	return bsb.nextLedger
+}
+
+// GetLedger returns the LedgerCloseMeta for the specified ledger sequence number
+func (bsb *BufferedStorageBackend) GetLedger(ctx context.Context, sequence uint32) (xdr.LedgerCloseMeta, error) {
+	bsb.bsBackendLock.RLock()
+	defer bsb.bsBackendLock.RUnlock()
+
+	if bsb.closed {
+		return xdr.LedgerCloseMeta{}, errors.New("BufferedStorageBackend is closed; cannot GetLedger")
+	}
+
+	if bsb.prepared == nil {
+		return xdr.LedgerCloseMeta{}, errors.New("session is not prepared, call PrepareRange first")
+	}
+
+	if sequence < bsb.ledgerBuffer.ledgerRange.from {
+		return xdr.LedgerCloseMeta{}, errors.New("requested sequence preceeds current LedgerRange")
+	}
+
+	if bsb.ledgerBuffer.ledgerRange.bounded {
+		if sequence > bsb.ledgerBuffer.ledgerRange.to {
+			return xdr.LedgerCloseMeta{}, errors.New("requested sequence beyond current LedgerRange")
+		}
+	}
+
+	if sequence > bsb.nextExpectedSequence() {
```

I think you also need to check that sequence must be >= bsb.lastLedger

---

Wouldn't this already cover the `>= bsb.lastLedger` case? Because if the `sequence is == lastLedger` or if `sequence == nextExpectedSequence()` the check should pass?

---

right but I'm thinking of the scenario where you prepare range starting from ledger 5 and then you request ledgers 5, 6, 7, and then 5 again. We should not allow you to go backwards so once you query for anything below the last ledger we should return an error